tests/input/inline_type_hint_assertions.php

```diff
@@ -0,0 +1,30 @@
+<?php
+
+declare(strict_types=1);
+
+final class Foo
+{
+    public function methodWithInlineHints()
+    {
+        /** @var Type $simpleType */
+        $simpleType = expression();
+
+        $typeDeclaredAfterExpression = expression();
+        /** @var Type $typeDeclaredAfterExpression */
+
+        $typeDeclaredViaAssertion = expression();
+        assert($typeDeclaredViaAssertion instanceof Type);
+
+        /** @var Type1|Type2 $unionType */
+        $unionType = expression();
+
+        /** @var Type1&Type2 $intersectionType */
```

Comment:
Unsure if we should deal with this now. I can see it growing in usage since it was first introduced a couple months ago, so it may be worth exploring after poking the @PHPStorm folks

Reply:
Issue for intersection types is pending and being ignored: https://youtrack.jetbrains.com/issue/WI-39419 for three months already. :(
Unfortunately until then, it's a no-go, it completely breaks assumptions in PhpStorm.

Reply:
cc @JanTvrdik

Reply:
Is that one actually a problem?

`assert($intersectionType instanceof Type1 && $intersectionType instacenof Type2);`

Reply:
The tools do not support type inference from `assert()`? Or complex assert expressions like this? I think, unlike intersection types which are not part of the standard, assertions as such are, so it's more likely to get the support of types inferred from assertions than from annotated intersection types.

Reply:
Tools don't know (yet) how to parse `Foo&Bar` in docblocks

Reply:
> Unsure if we should deal with this now.

I'm confused. As far as I understand, this ticket is about using assertions instead of annotations (including the intersection types). Why should the fact of us replacing them with assertions be defined by how annotations are supported by the tools?

Reply:
@morozov the CS tool simply needs to learn to parse them in order to replace them. These annotations are still valid in docblocks.

Reply:
@muglug awesome! Now we just need to convince JetBrains 😂

Reply:
> These annotations are still valid in docblocks.

Do you also want to replicate complex argument types with assertions? Otherwise, unlike local variables, they remain unenforced. For instance:
```php
/**
 * @param string|int $a
 */
functioon ($a) {
    assert(is_string($a) || is_int($a));
}
```

Reply:
Might be interesting, but unlikely for this PR

Reply:
I'm not too sure about how I feel about assertions for public API parameters. Personally of course I always avoid complex types in public APIs and only use them in private APIs when I want to keep things simple.

The thing with assertions is, that they are not ran by default (heck, PHP throws away the assertion op-codes when assertions are disabled). Thus most users wouldn't get those checks. IMHO a public API should do parameter checks the proper way (and throw exceptions itself) for such stuff.

Assertions come in handy when you have methods which are supposed to return something specific, but the API cannot specify the return type (like PSR containers). Thus you assert on the return type.